kafka-cron/consumer/main.go

```diff
@@ -0,0 +1,138 @@
+package main
+
+import (
+	"bytes"
+	"log"
+	"os"
+	"os/exec"
+	"os/signal"
+	"sync"
+
+	kingpin "gopkg.in/alecthomas/kingpin.v2"
+
+	"encoding/json"
+
+	"github.com/Shopify/sarama"
+	"github.com/berkeli/kafka-cron/types"
+	"github.com/google/uuid"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	brokerList = kingpin.Flag("brokerList", "List of brokers to connect").Default("localhost:9092").Strings()
+	topic      = kingpin.Flag("topic", "Topic name").Default("jobs-cluster-a").String()
+	retryTopic = kingpin.Flag("retryTopic", "Retry topic name").Default("jobs-cluster-a-retries").String()
+	partition  = kingpin.Flag("partition", "Partition number").Default("0").Int32()
+)
+
+func main() {
+
+	InitMonitoring(2112)
+
+	kingpin.Parse()
+	config := sarama.NewConfig()
+	config.Consumer.Return.Errors = true
+	brokers := *brokerList
+	master, err := sarama.NewConsumer(brokers, config)
+	if err != nil {
+		log.Panic(err)
+	}
+
+	config = sarama.NewConfig()
+	config.Producer.RequiredAcks = sarama.WaitForAll
+	config.Producer.Return.Successes = true
+	producer, err := sarama.NewSyncProducer(*brokerList, config)
+	if err != nil {
+		log.Panic(err)
+	}
+	defer func() {
+		if err := producer.Close(); err != nil {
+			log.Panic(err)
+		}
+	}()
+
+	defer func() {
+		if err := master.Close(); err != nil {
+			log.Panic(err)
+		}
+	}()
+	consumer, err := master.ConsumePartition(*topic, *partition, sarama.OffsetNewest)
+	if err != nil {
+		log.Panic(err)
+	}
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+	chDone := make(chan bool)
+	wgWorkers := sync.WaitGroup{}
+	go func() {
+		for {
+			select {
+			case err := <-consumer.Errors():
+				log.Println(err)
+			case msg := <-consumer.Messages():
+				var cmd types.Command
+				err := json.Unmarshal(msg.Value, &cmd)
+				if err != nil {
+					log.Println(err)
+				}
+				wgWorkers.Add(1)
+				// TODO: Add a worker pool with semaphore? What if jobs are dependent on each other?
+				processCommand(producer, cmd, &wgWorkers)
+			case <-signals:
+				chDone <- true
+				return
+			}
+		}
+	}()
+	<-chDone
+	log.Println("Interrupt is detected, shutting down gracefully...")
+	wgWorkers.Wait()
```

Comment:
pretty sure he's waiting for running workers to complete, but it could use a comment alright

Reply:
Yeah so I didn't want to terminate the program in case a command was already running.

Reply:
Can you talk through what is the series of events which could cause you to terminate the program while a job was running? Bearing in mind that once a single `case` in a `select` is picked, it will synchronously run until that `case` completes.

Reply:
So my thought process: 

1. We have a consumer(instance of an application) that picks up a job that runs for let's say 10 minutes. 
2. In the 5th minute the consumer receives a signal to terminate (this could be by a user or CI/CD tool that's upgrading/replacing the consumer)
3. I thought in such case it shouldn't just shut down, and it should wait for the last job to finish.

Reply:
Here's a small program which I think models your situation - before you try running it, try to reason through what it will print and do if you run it then send it a SIGINT.

```go
package main

import (
    "fmt"
    "os"
    "os/signal"
    "time"
)

func main() {
    ch := make(chan struct{}, 1)
    signals := make(chan os.Signal, 1)
    signal.Notify(signals, os.Interrupt)
    ch <- struct{}{}

    for {
        select {
        case <-signals:
            fmt.Println("Got signal - terminating 1")
            os.Exit(1)
        case <-ch:
            fmt.Println("Sleeping for 30 seconds then terminating 0")
            time.Sleep(30 * time.Second)
            os.Exit(0)
        }   
    }   
}
```

Once you've made your prediction, try running it - does it to what you expect?

Compare it with the structure of your actual program - what are the similarities or differences between your control flow and mine?

Reply:
After doing some testing, it didn't work as expected and I had to add the following option to cmd: 

```
cmd.SysProcAttr = &syscall.SysProcAttr{
		Setpgid: true,
	}
```
and run it in it's own goroutine.

Reply:
I see what you mean now, for some reason I tunnel visioned inside 1 case of the select and wasn't thinking about the first thing that happens when a signal is received, will have a think about the implementation

Reply:
What's specifically your goal? To wait for running processes to finish? Or to interrupt them and make sure they die before we do?